### PR TITLE
Organisation Units Sync

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

* **Issue:** Work #8 

### 💻 Status

- Do not merge, work in progress

### :memo: Implementation

- Add basic synchronization logic

- Allow accessing d2 models on-the-fly

### CODE REVIEW

File: src/components/sync-configurator/BaseSyncConfigurator.jsx [hunk 1]
```diff
@@ -22,8 +36,19 @@ export default class BaseSyncConfigurator extends React.Component {
         },
     ];
 
+    selectionChange = selection => {
+        this.setState({ selection });
+    };
+
+    startSynchronization = () => {
+        // TODO: Render new dialog to show summary and select destination instances
+        queryMetadata(this.props.d2, "organisationUnit", "zGv4gjlRmCX").then(result =>
```
Comment (SferaDev) [inline src/components/sync-configurator/BaseSyncConfigurator.jsx]: [DEBUG]: Should open the new synchronization dialog to show a summary of sync and select instances

File: src/logic/synchronization.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import { cleanObject, getAllReferences } from "../utils/d2";
+import { D2 } from "../types/d2";
+import {
+    NestedRules,
+    SynchronizationBuilder,
+    SynchronizationResult,
+} from "../types/synchronization";
+import { mergeCustomizer } from "../utils/lodash";
+
+function buildNestedRules(rules: string[]): NestedRules {
```
Comment (SferaDev) [inline src/logic/synchronization.ts]: Should be moved somewhere on utils

File: src/logic/synchronization.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+export async function queryMetadata(
+    d2: D2,
+    type: string,
+    ids: string | string[]
+): Promise<SynchronizationResult> {
+    const myClass = d2ModelFactory(d2, type);
+
+    // Fetch metadata API to export
+    const elements = _.isArray(ids) ? ids : [ids];
+    return await fetchMetadata(
```
Comment (SferaDev) [inline src/logic/synchronization.ts]: Same with object destructuring

File: src/logic/synchronization.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        d2,
+        type,
+        elements,
+        myClass.getExcludeRules(),
+        myClass.getIncludeRules()
+    );
+}
+
+export async function startSynchronization(d2: D2, builder: SynchronizationBuilder): Promise<void> {
```
Comment (SferaDev) [inline src/logic/synchronization.ts]: Missing origin/destination instance logic

File: src/models/d2Model.ts [hunk 5]
```diff
@@ -47,17 +63,43 @@ abstract class D2Model {
 }
 
 export class OrganisationUnitModel extends D2Model {
-    protected static metadataType = 'organisationUnit';
+    protected static metadataType = "organisationUnit";
+    protected static excludeRules = [
+        "legendSets",
+        "user",
```
Comment (SferaDev) [inline src/models/d2Model.ts]: All the user related clean-up could be on by default for all models

File: src/utils/lodash.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import _ from "lodash";
+
+export const mergeCustomizer = (obj: any, src: any): any =>
```
Comment (SferaDev) [inline src/utils/lodash.ts]: Should be a mixin but I was having issues getting mixins to work on TS

File: src/logic/synchronization.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        },
+        {}
+    );
+}
+
+async function get(d2: D2, elements: string[]): Promise<any> {
+    let promises = [];
+    for (let i = 0; i < elements.length; i += 100) {
+        let requestUrl =
+            d2.Api.getApi().baseUrl +
```
Comment (SferaDev) [inline src/logic/synchronization.ts]: We should read the origin instance baseUrl instead through parameter

File: src/utils/lodash.ts (hunk 6 above)
Comment (tokland) [reply src/utils/lodash.ts]: I already fought with this, check:

https://github.com/EyeSeeTea/vaccination-app/blob/development/src/utils/lodash-mixins.ts

Comment (SferaDev) [reply src/utils/lodash.ts]: Perfect! Thanks :)

**Pull request closed without merging**